ports/stm32/subghz.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2023 Angus Gratton
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "subghz.h"
+#include "irq.h"
+#include "spi.h"
+
+#if defined(STM32WL)
+
+// Interface to the STM32WL series "SUBGHZ Radio" module
+
+STATIC void handle_radio_irq() {
+    // Level-triggered interrupts means the interrupt has to be cleared before
+    // this function returns.
+    //
+    // Rather than writing to SUBGHZ SPI in Interrupt Context to clear the
+    // interrupt, disable the IRQ and rely on Python code to call
+    // subghz_irq(handler) to re-enable when needed.
+    HAL_NVIC_DisableIRQ(SUBGHZ_Radio_IRQn);
+
+    mp_obj_t callback = MP_STATE_PORT(subghz_callback);
+    if (callback != mp_const_none) {
+        // Passing dummy 'pin' argument of None, to keep
+        // compatibility with machine.Pin.isr() handlers.
+        mp_call_function_1(callback, mp_const_none);
```

Comment:
since this is a hard IRQ, it needs further protection:
```c
mp_sched_lock();
gc_lock();
mp_call_function_1_protected(...);
gc_unlock();
mp_sched_unlock();
```

Reply:
:facepalm: I switched this at the last minute, clearly without thinking! Thx.